### DISCUSSION TITLE
What is the CsWin32 way to handle API failures and access error information?

### DISCUSSION THREAD

**Discussion by netcorefan1** [microsoft/CsWin32#1092] [Q&A] [answered]
The title says it all. 
```
if (ret.IsNull) {
   var err = new Win32Exception(Marshal.GetLastWin32Error())
}
```
The above is not reliable and sometime returns `The operation completed successfully.` even on failures.
Does CsWin32 provides some implementarion for error handling?

**Comment by AArnott** [comment, marked as answer]
CsWin32 generates APIs that follow the same Win32 error reporting standard that you might have in manually written P/Invokes. You should refer to the documentation of the individual win32 API to learn how to get error information. _Some_ APIs call SetLastError while others do not. You should _only_ call `GetLastWin32Error` directly after a call to a win32 API that is documented as calling `SetLastError` and should only do so when it says you should (e.g. when you get a null result).
The reason for this is if you call an API that does _not_ call SetLastError (either because it never does, or only does in error cases), then by calling `GetLastWin32Error` you'll be getting the code from some prior call to SetLastError that may be unrelated to the function you just invoked.

**Reply by netcorefan1** [reply]
Many thanks for your help. That's all I needed to know. It should be straightforward to adapt my old error manager handler for CsWin32. Just wanted to be sure there was no `CsWin32ErrorManager` available (which would have been for sure a better choice).
Thanks for reminding me about `SetLastError`!
Thanks

**Comment by 0x5bfa** [comment]
Is there enum for Win32 erro like Win32Error.ERROR_ACCESS_DENIED?

**Reply by AArnott** [reply]
Yes. The enum is called `WIN32_ERROR`.
